### ISSUE TITLE
Hard coded Grid Names and grid specific references 

### ISSUE THREAD

**Issue by Vinhold** [InWorldz/halcyon#50]
InWorldz shows up in the viewer in several places and the money amount shows as $Iz or sometimes as $Os. It is not good maintenance to have multiple places to set the grid name or the common Library folder name or any other grid specific title that will require changing every hard coded place they occur or trying to use the region settings to do it. 

Regardless of how a process may run and what information it may be able to access from the Halcyon ini file, all processes should have access to the database system. If the process supplies some display data that is grid specific it could get that from the database. If nothing else, there should be one place where grid information can be set up and all needed processes can get it once on startup and use it there after. Once set for a grid it will not likely be changing very often. I think this will require some careful consideration before it becomes a bigger problem with many dependent grids.

**Comment by jimtarber** [comment]
I thought this info was returned in the economy data but that is just the numbers. But it is part of the login response as the "currency" member of the login response.  So this could be provided. I did however find many many hard-coded references to "L$", but that is a viewer issue if it is not using the string specified by "currency".

So a start would be to specify something in the "currency" member of the login response, however the other question is where the data would come from. I'm thinking it probably belongs right beside the grid info, perhaps second block of data separate from the viewer's existing GridInfo data, perhaps like a GridConfig block. I don't think it needs to support dynamic changes, so reading it at server startup would be fine. I believe this response comes from the login user server, so it would already have the data.

**Comment by Vinhold** [comment]
I agree. While I am not sure of where all of the viewer display content is provided that is grid specific, it would seem logical that it should be supplied when a viewer connected to the grid. The [GridInfo] block provides all the grid URL services and the gridname and gridnick entries. You are suggesting that any other grid specific definitions would be supplied in a new block like this?
[GridConfig]
; Set currency label for the grid. Default is OS.
currency = "Iz"
; Set grid Library name prefix. Default is no name set. ie: Library. Providing a name would make it InWorldz Library.
libraryname = "InWorldz"
....
and so on for any other grid specific viewer text?
I like that idea or if the number of entries is small like the two listed above, they could be added to the [GridInfo] block.

**Comment by kf6kjg** [comment]
I actually made a list of the various other non-currency hard-coded name references and have a starting patch in place: just need to get it disentangled from some other orthogonal changes...  Some places like the common library being "InWorldz Library", and a few other places that would need to request the info from the owner of said info, are not covered yet as that's a larger change.

**Comment by jimtarber** [comment]
Be very careful changing the name of the Library, as viewers (at least IW3) test for specific names in some cases. e.g. I think it only specifically checks for "InWorldz Library" and "My Library" but I forget the specifics.

**Comment by Vinhold** [comment]
Actually the default name "Library" would be fine by me if no other name is defined. If the viewers are looking for and testing for some custom name that is very nice of them. But they should just take the name given and use the folder type to determine its function as the Library folder or a combination of containing the name "Library" and the folder type. That way it could be given a custom prefix name and does not break anything. My purpose in providing a custom name entry for the Library was to allow the trend set by OpenSim and InWorldz of adding a prefix name to the Library, but just have it set in the Halcyon.ini if it is to be used or leave blank for the default name: Library.

The problem I have at the moment is the Library in Gospel World shows as "InWorldz Library". While I love and appreciate InWorldz, having the name showing in the Gospel World grid is just a bit confusing and inaccurate. :)

**Comment by jimtarber** [comment]
I think there's another safe name you can use (it may be "My Library" but I can't confirm that without searching the viewer source).  The reason I commented there was to alert that it mattered to some viewers; it wasn't arbitrary text.

**Comment by kf6kjg** [comment]
Just grep'd IW3 source code.  There are no instances of "My Library" or "InWorldz Library" - in fact, I can find not special handling of anything that has "Library" in the name.  Maybe earlier viewers had such a hack?  It does handle the library specially, but it doesn't appear to handle it by name checks.

Even a search for "InWorldz" didn't turn up anything interesting.

I did find some special handling for "My Inventory" however: `void LLInventoryPanel::openStartFolderOrMyInventory()` does a does a linear search of the direct children of the inventory root and opens the first item named "My Inventory".

**Comment by jimtarber** [comment]
It's not in the viewer source _code_; it's in the viewer data files.  

It doesn't work like that in the viewer; you can't just search for string constants. Many of the string constants (including the library root folder name) come from the ```strings.xml`` file in the viewer *skin*, so it depends on the viewer, depends on the skin the user has selected, which also means it is localized to different languages, etc.

And beyond that, the list of viewers is pretty long, some aren't even based on the LL C++ code. Checking the LL-based source code is not enough for a test of backwards-compatibility.  (I think Lumiya, for example, the Android viewer, is written in original Java code.  Pocket Metaverse on iOS probably isn't base don the LL code. I don't know if Stored Inventory still works either, that's probably just a modified OMV viewer.)

But we know the existing names already in use are safe so since SL uses "Library", an IW-specific name could be replaced with the generic one, and probably should be. (Whoever changed it to "InWorldz Library" took an even larger risk of breaking viewers.)  Let's get the generic SL names back in place.

**Comment by jimtarber** [comment]
Also, in this case, I think I remember what I was thinking of above and that actually applied to "My _Inventory_", not "My _Library_", where there was code in some viewers to recognize that one by name.  

So I think the specific case I was worried about doesn't apply here if you're talking about only changing the library folder only.

Regardless, none of this is server-side. The name isn't something we can configure server-end because it doesn't come from the server, it comes from the viewer's skin. If you want to change the library root folder name, it's a _viewer_ change.

**Comment by kf6kjg** [comment]
The library name is being set and sent by the region as it stands.  I understand that the skins can change the name string, and that they do so by matching strings, so I agree that it should be "Library" - however since IW already long-since changed that, the code I just put together this evening leaves "InWorldz Library" as the default to remain compatible with the existing grid configuration, but the grid owner can then override it to whatever they want.  I gave the sample INI file this note in the [StandAlone] section:
```
    ; Sets the name of the library folder that is common to all users.
    ; Overridden by the User Server when running in grid mode. If you are
    ;  running in grid mode, please set the library_name attribute in UserServer_Config.xml
    ; WARNING: Using any other value than "Library" is going to cause language
    ;  translation issues in the viewers.
    ; Default is "InWorldz Library"
    LibraryName = "Library"
```

This way it can be brought back to sanity or not at the whim of the grid monkey.

**Comment by kf6kjg** [comment]
You can monitor and comment on the changes over at https://github.com/kf6kjg/halcyon/commits/branding_scrub if you wish.

**Comment by jimtarber** [comment]
I think we all agree that it is important for the Halcyon project to remove the hard-code names for grid-specific texts, like "L$", and "InWorldz" strings, but I don't know what the goal is on changing the Library name to anything other than "Library"? I don't even think it is returned to the viewer, the viewer uses a separate requests for Library folder contents, and starts with the root folder: FetchLib, FetchLib2, FetchLibDescendents and FetchLibDescendents2.  It then uses strings from the language-specific skin to display this tree.  What is the purpose of having the Library name configurable, at the server end?

**Comment by kf6kjg** [comment]
It is in fact returned by the User Server in grid mode, and returned by the region server in StandAlone.  I've verified that with the code I committed last night to the above linked branch, that the viewer (Alchemy) does indeed show the custom string.  However it's not likely that any other string than "Library" will be successfully translated to other languages as the comparison would fail.

The sole reason I made it configurable was to not cause a customer support nightmare on IW.  I also made "InWorldz Library" the default for the very same reason.  If the string were to suddenly change all the support teams would need to know that it has changed, any screenshots or other documentation showing how to access it would need changing, etc.  Grids coming from OS will likely have "OS Library" IIRC as their old default and may want to keep that or rebrand - as long as they know that it'll cause a translation issue, it's their choice.

Main thing: I've been working hard to make my changes not require any "administrative" choices other than checking to see if the code is clean.  Also means that no one needs to, although it's a good idea, dig into the the source for the support tools and add the new options: the existing configurations will do the right thing for the IW grid.

**Comment by jimtarber** [comment]
Okay that sounds like good reasons; and I appreciate that you're trying to keep things compatible with the current behavior. If Alchemy is displaying the custom name, then allowing the customization does serve a purpose, and it's a good change in that sense.

But I think this is one of those cases where none of the grids would want it to be a non-standard name anyway. Does IW need to label it "InWorldz Library"? Or GW name it "GW Library" or something?  I think it's one of those things where you have looked into it now and together we've identified the behavior, and if the viewer needs 'Library' in order to recognize (and translate the name), then that's what the server should give it so that the viewer code just works regardless of the grid.

I think the existing InWorldz customization is really just a viewer bug caused by an overzealous branding attempt, and it should be fixed (back to 'Library').  I don't think the name of the library there is a problem for our support folks (there are only 2) and it's more likely that there are instructional videos recorded on SL anyway, and so if we're straying from that, we should fix it.  

Now this change does allow us to set it to whatever we want, but it requires a reconfig if the default remains something other than 'Library'.  That might be a good approach in the short term; require a change to our templates to restore the standard name and if that causes any grief, we can just reconfig it back the way it was.  But doesn't quite seem right with the default being a string that causes the viewer to be unable to localize it (if that's the case).

So to summarize this, I think we should just fix the somewhat bogus IW library name and use the generic one.  Making it configurable so that others can use the generic name and IW can keep a problematic one is adding new complexity to preserve a IW bug; so I think it's going in the wrong direction and we should just fix the 'Library' naming.  Let viewers customize it in the skin as designed.

**Comment by Vinhold** [comment]
Great progress on this in 0.9.34 release! only a problem which I think Ricky added a fix to for the currency_symbol entry will make all posted here to be working.

**Comment by Vinhold** [comment]
All changes discussed here have been applied in 0.9.36 Halcyon release. L$ is settable in Halcyon.ini [Economy]
